cmd/gateway/zerostor/meta/filesystem_meta.go

```diff
@@ -0,0 +1,1059 @@
+	}
+	return objMeta, m.WriteObjMeta(&objMeta)
+}
+
+func (m *filesystemMeta) scan(ctx context.Context, bucket, prefix, after string, delimited bool, maxKeys int, result *minio.ListObjectsV2Info) (string, error) {
+	log.WithFields(log.Fields{
+		"bucket":    bucket,
+		"prefix":    prefix,
+		"after":     after,
+		"delimited": delimited,
+		"maxKeys":   maxKeys,
+	}).Debug("scan bucket")
+
+	root := filepath.Join(m.objDir, bucket)
+
+	var prefixed string
+	if delimited {
+		//so prefixed should be a directory
+		prefixed = filepath.Join(root, prefix)
+	} else {
+		//otherwise we search the root
+		prefixed = root
+	}
+
+	var last string
+
+	stat, err := os.Stat(prefixed)
+	if os.IsNotExist(err) {
+		return "", nil
+	} else if err != nil {
+		return "", err
+	}
+
+	if !stat.IsDir() {
+		//is a file
+		md, decodeErr := m.decodeObjMeta(m.objectFile(bucket, prefix))
+		if decodeErr != nil {
+			return "", decodeErr
+		}
+
+		result.Objects = append(result.Objects, createObjectInfo(bucket, prefix, &md))
+		return "", nil
+	}
+
+	err = filepath.Walk(prefixed, func(path string, info os.FileInfo, err error) error {
+		if prefixed == path {
+			return nil
+		}
+
+		if err != nil {
+			return errors.Wrapf(err, "failed to walk '%s'", prefixed)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		if maxKeys == 0 {
+			return errMaxKeyReached
+		}
+
+		name := strings.TrimLeft(strings.TrimPrefix(path, root), "/")
+		if strings.Compare(name, after) <= 0 {
+			//scanning until reach the "after"
+			return nil
+		}
+
+		maxKeys--
+		if !delimited && !strings.HasPrefix(name, prefix) {
+			//match on full prefix and the full object name doesn't have this file/dir
+			return nil
+		}
+
+		last = name
+		if delimited && info.IsDir() {
+			result.Prefixes = append(result.Prefixes, filepath.Clean(name)+"/")
+			return filepath.SkipDir
+		}
+
+		if info.IsDir() {
+			//flat listing (no delmeter)
+			result.Objects = append(result.Objects, minio.ObjectInfo{
+				Bucket: bucket,
+				Name:   filepath.Clean(name) + "/",
+				ETag:   name,
+				IsDir:  true,
+			})
+
+			return nil
+		}
+
+		//we found a file
+		// if file, get metadata of this file
+
+		md, err := m.decodeObjMeta(m.objectFile(bucket, name))
+		if err != nil {
+			return err
+		}
+
+		result.Objects = append(result.Objects, createObjectInfo(bucket, name, &md))
+		return nil
+	})
+
+	return last, err
+}
+
+func (m *filesystemMeta) decodeObjMeta(file string) (ObjectMeta, error) {
+	// open file
+	f, err := os.Open(file)
+	if os.IsNotExist(err) {
+		return ObjectMeta{}, minio.ObjectNotFound{}
+	} else if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"subsystem": "disk",
+			// "key":       string(key),
+		}).Error("failed to get metadata")
+		return ObjectMeta{}, err
+	}
+
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return ObjectMeta{}, err
+	}
+	if !stat.IsDir() {
+		objMeta := new(ObjectMeta)
+		objMeta.Metadata = *new(metatypes.Metadata)
```

Comment:
I really don't get this code here!!
U don't need to new a `value` of struct .. ever, it will always be there but with empty fields (zero value of each) 
so objMeta := new(ObjectMeta) is enough to make a full value of objMeta, including objMeta.Metadata . Unless of course Metdata is itself a pointer, in that case a call to new is needed, but that's not the case here.

Reply:
it used to be a pointer, hence I have a * before the new. But seems I forgot this part after I changed the struct.